### PR TITLE
Implement final FilMultihash/FilCodec

### PR DESCRIPTION
**Pull request by ribasushi**
( I do not have fil-project perms, hence PR from fork )

This rewrites the module rather radically, implementing the decision agreed by @rvagg, @jbenet, @hannahhoward and @stebalien in https://github.com/multiformats/multicodec/pull/161#issuecomment-618160880

Focus on the function name changes and signature changes within the module: the tests are simply adjusted to match.

### CODE REVIEW

**Comment by rvagg** [comment]
Duplication is a thing I'm mostly interested and maybe we can reduce that a bit? Starting with the constants.

Sealed and unsealed codecs are defined here:

https://github.com/ipfs/go-cid/blob/85cd30874e614c61d6dd82d991878d84053fd579/cid.go#L78-L79

The two hash functions are defined here (I think only the sha256 one is used here at the moment):

https://github.com/multiformats/go-multihash/blob/6f1ea18f1da5f7735ea31b5e2011da61c409e37f/multihash.go#L71-L73

A lot of the code in this repo exists because go-cid and go-multihash didn't have values ready, so a further duplication could happen by pushing all of the multihash and cid creation back down to go-cid and go-multihash now that they have them. But I think that's lower priority since the main urgent thing is simply getting the right constants into play! Up to you how far down that rabbit hole you go I think, it can be deferred. For now I'd like to see the constants deduplicated and not redefined if that's not too hard. The "reserved" entries are probably in the way a bit, but I don't think they're used at all and I'm not really sure why they're there. The `0xf100` space is essentially "reserved" now for filecoin.

**Comment by ribasushi** [comment]
Implements consensus from https://github.com/multiformats/multicodec/pull/161#issuecomment-618160880
This is now open for final review by the original stakeholders:  @rvagg, @jbenet, @hannahhoward @Stebalien and @mikeal. ( Can't assign reviewers due to lack of perms )

This will need to move relatively fast to make it into the feature freeze. I can help with updating lotus and friends, provided the changes are 100% correct / approved.

File: commcid.go [hunk 1]
```diff
@@ -129,18 +83,18 @@ func CIDToDataCommitmentV1(c cid.Cid) ([]byte, error) {
...
 
 // ReplicaCommitmentV1ToCID converts a raw data commitment to a CID
 // by adding:
-// - serialization type of raw
-// - hashing type of Filecoin sealed hashing function v1 (0xfc2)
+// - codec of type FC_SEALED_V1
+// - hashing type of FC_SHA2_256_TRUNC254
 func ReplicaCommitmentV1ToCID(commR []byte) cid.Cid {
-	c, _ := CommitmentToCID(commR, FC_SEALED_V1)
+	c, _ := CommitmentToCID(commR, cid.FilCommitmentSealed, multihash.SHA2_256_TRUNC254_PADDED)
```
Comment (rvagg) [inline commcid.go:97]: iirc CommR uses sealed+Poseidon. CommD and CommP use unsealed + the special sha256. @porcuquine can you confirm please before we commit to these identifiers?

Comment (porcuquine) [reply commcid.go:97]: That's right. CommR is the output of a binary Poseidon hash and has no relationship to SHA256. Does that suffice to answer your question?

File: commcid.go [hunk 2]
```diff
@@ -10,115 +10,69 @@ import (
...
 	}
-	return decoded.Digest, code, nil
+	return decoded.Digest, FilMultiCodec(c.Type()), nil
 }
 
 // DataCommitmentV1ToCID converts a raw data commitment to a CID
 // by adding:
-// - serialization type of raw
-// - hashing type of Filecoin unsealed hashing function v1 (0xfc2)
+// - codec of type FC_UNSEALED_V1
```
Comment (rvagg) [inline commcid.go:72]: `codec of type FilCommitmentUnsealed`

File: commcid.go [hunk 3]
```diff
@@ -129,18 +83,18 @@ func CIDToDataCommitmentV1(c cid.Cid) ([]byte, error) {
...
 		return nil, ErrIncorrectHash
 	}
 	return commD, nil
 }
 
 // ReplicaCommitmentV1ToCID converts a raw data commitment to a CID
 // by adding:
-// - serialization type of raw
-// - hashing type of Filecoin sealed hashing function v1 (0xfc2)
+// - codec of type FC_SEALED_V1
```
Comment (rvagg) [inline commcid.go:94]: `codec of type FilCommitmentSealed`

File: commcid.go (hunk 1 above)
Comment (rvagg) [reply commcid.go:97]: @porcuquine yep, thanks!

@ribasushi so, better change this to `multihash. POSEIDON_BLS12_381_A1_FC1` and the comment above it too

File: commcid.go [hunk 4]
```diff
@@ -172,18 +126,14 @@ func CIDToPieceCommitmentV1(c cid.Cid) ([]byte, error) {
...
+	return mc == cid.FilCommitmentUnsealed || mc == cid.FilCommitmentSealed
 }
 
 // ValidFilecoinMultihash returns true if the given multihash type
 // is recognized as belonging to filecoin
-func ValidFilecoinMultihash(code FilecoinMultihashCode) bool {
-	_, ok := FilecoinMultihashNames[code]
-	return ok
+func ValidFilecoinMultihash(mh uint64) bool {
+	return mh == multihash.SHA2_256_TRUNC254_PADDED // sha2-256-trunc2 is all we support for now
```
Comment (rvagg) [inline commcid.go:138]: `|| mh == multihash.POSEIDON_BLS12_381_A1_FC1`

File: commcid_test.go [hunk 5]
```diff
@@ -77,11 +77,11 @@ func TestReplicaCommitmentToCID(t *testing.T) {
 
 	c := commcid.ReplicaCommitmentV1ToCID(randBytes)
 
-	require.Equal(t, c.Prefix().Codec, uint64(cid.Raw))
+	require.Equal(t, c.Prefix().Codec, uint64(cid.FilCommitmentSealed))
 	mh := c.Hash()
 	decoded, err := multihash.Decode([]byte(mh))
 	require.NoError(t, err)
-	require.Equal(t, decoded.Code, uint64(uint64(commcid.FC_SEALED_V1)))
+	require.Equal(t, decoded.Code, uint64(multihash.SHA2_256_TRUNC254_PADDED))
```
Comment (rvagg) [inline commcid_test.go:84]: `multihash.POSEIDON_BLS12_381_A1_FC1`

File: commcid_test.go [hunk 6]
```diff
@@ -92,10 +92,10 @@ func TestCIDToReplicaCommitment(t *testing.T) {
 	require.NoError(t, err)
 
 	t.Run("with correct hash format", func(t *testing.T) {
-		hash := testMultiHash(uint64(commcid.FC_SEALED_V1), randBytes, 0)
+		hash := testMultiHash(multihash.SHA2_256_TRUNC254_PADDED, randBytes, 0)
```
Comment (rvagg) [inline commcid_test.go:95]: `multihash.POSEIDON_BLS12_381_A1_FC1`

**Comment by rvagg** [comment]
It's certainly simplified after this! maybe this is a style issue - but would it be easier if this package re-exported those go-multihash and go-cid constants so callers don't need to reach into those packages to get them? The current usage of this package would be using constants from here but will not need to include this package, go-cid and go-multihash to get everything together. Maybe that's perfectly fine for Go and nobody will blink an eye at it, it just seems heavy-handed.

**Comment by rvagg** [comment]
Maybe `DataCommitmentV1ToCID()` and `ReplicaCommitmentV1ToCID()` are the main interfaces here, in which case my comment about constants may not be that relevant. The other two public methods exposed here require you supply the codec and multihash constants to use, so you need to get them from somewhere. Prior to this PR you'd just pull them in from this package. By removing them now and deferring to go-cid and go-multihash, if you call those methods then you'd also need to grab the relevant multihash and multicodecs from go-cid and go-multihash.

File: commcid.go (hunk 1 above)
Comment (ribasushi) [reply commcid.go]: Offfff thanks for catching this folks! I misread an earlier comment to arrive at this 🤦 
All fixed!

Also rewrote the validator from scratch to disallow incorrect codec/hash combinations. Namely:
https://github.com/filecoin-project/go-fil-commcid/blob/226ec33a83e9c5cd6d01029cbff44590e70fe92c/commcid.go#L111-L133

**Comment by ribasushi** [comment]
> Maybe DataCommitmentV1ToCID() and ReplicaCommitmentV1ToCID() are the main interfaces here, in which case my comment about constants may not be that relevant. 

@rvagg right, this is the idea, cemented further by the validator. I am inclined to leaving the "inconvenience" in place at present, so that folks do not reach out for these functions without thinking. **Ideally** none of the helpers should be exported, but I do not want to change the preexisting package's style *too* much.

**Comment by ribasushi** [comment]
Last weeks update on this was:

> we can consume this into our staging branch as soon as it’s merged (all consensus-critical stuff goes there).

@arajasek @whyrusleeping what needs to happen to make this... happen? :)

**Comment by rvagg** [comment]
@hannahhoward I think you're the closest to being an "owner" here, would you mind taking a look and merging?

**Comment by mikeal** [comment]
@whyrusleeping does there need to be some kind of plan to roll this out to lotus? it’s sort of a breaking change, but you’d know better than me what scale of breakage it could potentially cause.

**Pull request merged**